hat/app/org/hatdex/hat/api/controllers/RichData.scala

```diff
@@ -70,342 +103,622 @@ class RichData @Inject() (
+    data.map(d => Ok(Json.toJson(d)))
+  }
 
-    def bundleNotFound(bundleId: String) = ErrorMessage("Bundle Not Found", s"Bundle $bundleId not found")
+  def getContractData(
+    namespace: String,
+    endpoint: String,
+    orderBy: Option[String],
+    ordering: Option[String],
+    skip: Option[Int],
+    take: Option[Int]): Action[AnyContent] =
+    UserAwareAction.async { implicit request => //    SecuredAction(
+      {
+
+        val ret = for {
+          isOk <- isValidContractRequest(request.body.asJson)
+          ret <- if (isOk)
+            makeData(namespace, endpoint, orderBy, ordering, skip, take)
+          else Future.successful(BadRequest)
+        } yield ret
+
+        ret
+      }
+    }
 
-    def dataUpdateMissing(message: String) = ErrorMessage("Data Missing", s"Could not update records: $message")
-    def dataDeleteMissing(message: String) = ErrorMessage("Data Missing", s"Could not delete records: $message")
-    def dataLinkMissing(message: String) = ErrorMessage("Data Missing", s"Could not link records: $message")
+  private def bodyOptToContractDataRequest(
+    bodyOpt: Option[JsValue]): Option[ContractDataRequest] = {
+    for {
+      bodyAsJson <- bodyOpt
+      token <- refineV[NonEmpty]((bodyAsJson \ "token").as[String]).toOption
+      hatName <- refineV[NonEmpty]((bodyAsJson \ "hatName").as[String]).toOption
+      contractId <- refineV[NonEmpty]((bodyAsJson \ "contractId").as[String]).toOption
+      cdr = ContractDataRequest(
+        ShortLivedToken(token),
+        HatName(hatName),
+        ContractId(java.util.UUID.fromString(contractId)))
+    } yield cdr
+  }
 
-    def dataCombinatorNotFound(combinator: String) = ErrorMessage("Combinator Not Found", s"Combinator $combinator not found")
+  // TODO: Use KeyId
+  private def requestKeyId(
+    contractDataRequest: ContractDataRequest): Option[String] = {
+    for {
+      keyId <- ShortLivedTokenOps
+        .getKeyId(contractDataRequest.token.value)
+        .toOption
+    } yield keyId
+  }
 
-    def richDataDuplicate(error: Throwable) = ErrorMessage("Bad Request", s"Duplicate data - ${error.getMessage}")
-    def richDataError(error: Throwable) = ErrorMessage("Bad Request", s"Could not insert data - ${error.getMessage}")
-    def forbidden(error: Throwable) = ErrorMessage("Forbidden", s"Access Denied - ${error.getMessage}")
-    def forbidden(message: String) = ErrorMessage("Forbidden", s"Access Denied - ${message}")
+  def isValidContractRequest(bodyOpt: Option[JsValue]): Future[Boolean] = {
+    val ret = for {
+      cdr <- bodyOptToContractDataRequest(bodyOpt)
+      keyId <- requestKeyId(cdr)
+    } yield (cdr, keyId)
+
+    ret match {
+      case Some((cdr, keyId)) => {
+        verifyTokenWithAdjudicator(cdr, keyId)
+      }
+      case _ => Future.successful(false)
+    }
   }
-}
 
+  // TODO: String to KeyId
+  def verifyTokenWithAdjudicator(
+    cdr: ContractDataRequest,
+    keyId: String): Future[Boolean] = {
```

Comment:
Discussion topic
Bool return type may be difficult to debug
If we attempt to validate a token and it fails how would we discover the reason for it failing ?
Suggest Either[VerificationADT, Result] this has several benefits
1. if verification ADT extends from ErrorADT and all functions return an Future[Either[ErrorADT, _]] then they can be flatmapped and more combined easily (you could remove the nested match statements and have a nest for comprehension (EitherT from cats/scalaz would remove the nesting)
2. returning a ADT allows more specific error information to be used for debugging or upstream actions (i.e. http status code)

Reply:
I am a fan of more specific error messages

Reply:
It's an Error ADT now